### ISSUE TITLE
TFLu wrong predictions for optimized model

### ISSUE THREAD

**Issue by lheim** [tensorflow/tensorflow#40866]
@tensorflow/micro

**System information**
- Host OS Platform and Distribution: macOS 10.14.16 and Ubuntu 18.04.4 LTS
- TensorFlow installed from: source
- Tensorflow version: TF v2.2
- Target platform: ARM Mbed OS (bare-metal) on STM32L4
- MCU Compiler: GCC ARM (GNU Arm Embedded Toolchain 9-2020-q2-update)

**Describe the problem**

Running an optimized model (see code snippet for generation below) on the MCU results in wrong predictions. Those predictions differ depending on the build profile (see more below).
Running the model locally with the TF Lite Interpreter works flawlessly.

**Please provide the exact sequence of commands/steps when you ran into the problem**

I'm currently benchmarking multiple models on the STM32L4 and want to verify the accuracy on the MCU itself by sending the test data and the results back and forth via UART.
This works flawlessly for the converted model without any optimizations:
```python
converter = tf.lite.TFLiteConverter.from_keras_model(tf_model)
tflite_model = converter.convert()
open('./TFLite-model/LeNet-MNIST.tflite', 'wb').write(tflite_model)
```

This also works flawlessly for the model with full int8 quantization:
```python
converter = tf.lite.TFLiteConverter.from_keras_model(tf_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.representative_dataset = yield_representative_dataset
tflite_model = converter.convert()
open('./TFLite-model/LeNET-MNIST_int8ops.tflite', 'wb').write(tflite_model)
```

However, the predictions are wrong when only using `[tf.lite.Optimize.DEFAULT]` - where only some of the weights are quantized.
```python
converter = tf.lite.TFLiteConverter.from_keras_model(tf_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
tflite_model = converter.convert()
open('./TFLite-model/LeNet-MNIST_optimized.tflite', 'wb').write(tflite_model)
```

Everything works flawlessly on the local TFLite Interpreter, and the problem only occurs with the mentioned model **with and without cmsis-nn** on the MCU.

Building with the `release.json` profile (mainly `-Os`) always results in the following prediction independent from the input:
```
[0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0]
```

However, building with the `debug.json` profile (`-Og`) always results in different wrong outputs but reproducible with the same input e.g.:
```
[0.1012007445, 0.1061911806, 0.0979489163, 0.1032304987, 0.1272625774, 0.0807816312, 0.0818515792, 0.0830543563, 0.0847685188, 0.1337099075]
```

**Things I already tried**
- the model works flawlessly on the TF Lite interpreter on my host machines
- the model is correctly ported to the MCU
    - no TFLu runtime errors
    - verified the input and output tensors size and datatypes (as expected float32)
- there is ongoing computation on the MCU and I can even benchmark each layer individually

I currently don't have a second MCU at hand to test this.

---

Given that all the other models run flawlessly on the MCU and that the output depends on the compiler optimizations, I'm a bit clueless and expect some problem with TFLu.

The only difference from the optimized to the non-optimized model are the quantized weights in the fully-connected layers, I expect a problem somewhere there.
Furthermore the fully int8 quantized model runs flawlessly.
So I would restricte the error to kernels which use the int8 quantized weights, but still do the computation in float32 -- this is the only spot where I do see a difference of the model architectures.

Thankful for any feedback and pointers.

**Attachements**
[GDrive Folder](https://drive.google.com/drive/folders/1bXUNiN5vrr1n1mrFuKsB0_h_nLqd0DD5?usp=sharing)
- [model architecture](https://drive.google.com/file/d/1UKt578IcToaQJdPrIVAvRexwb6XEyJBd/view?usp=sharing)
- the original keras model
- the converted optimized model which gives wrong predictions

**Comment by petewarden** [comment]
Thanks for the report, and sorry you're hitting problems!

This definitely seems like a bug, but the behavior that I would expect here is that the kernels would report an error because they don't support the combination of float calculations and eight-bit weights. This isn't well documented, but for code size reasons we focus on either pure eight-bit activations and weights, or float activations and weights, and don't support the "hybrid" combination of float activations and eight-bit weights in TFL Micro. This combination is supported in the mobile version of TFL, which is why it works in the Python interpreter, since that's using the mobile version.

We should report an error here rather than letting the calculations continue, we should figure out which kernel is failing silently and allowing uninitialized data to flow through the graph.

To solve your problem, can we help you figure out how to convert your graph to fully quantized weights and activations? Are you seeing accuracy issues with that approach right now?

**Comment by lheim** [comment]
Thanks for getting back to me @petewarden.

This makes sense and confirms my observations. The *pure* models with int8 or either float32 activations **and** weights do work flawlessly.

I pasted my code snippet for a graph with fully quantized weights and activations above:
```python
converter = tf.lite.TFLiteConverter.from_keras_model(tf_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.representative_dataset = yield_representative_dataset
tflite_model = converter.convert()
open('./TFLite-model/LeNET-MNIST_int8ops.tflite', 'wb').write(tflite_model)
```

The resulting model works and performs well, especially using cmsis-nn. 😊 

I'm fine with either closing the issue or, as you mentioned, keep it open to track the work on an error report for the kernel. In my case the layer is probably the `FULLY_CONNECTED` layer.

**Comment by cloudwiser** [comment]
And for anyone running TF 1.15 the tweaked conversion script below (derived from TinyML by @petewarden) also fully quantizes your h5 model...

`converter = tf.lite.TFLiteConverter.from_keras_model_file('LeNet-MNIST.h5')`
`converter.inference_input_type = tf.lite.constants.INT8`
`converter.inference_output_type = tf.lite.constants.INT8`
`converter.optimizations = [tf.lite.Optimize.DEFAULT]`

**Comment by sushreebarsa** [comment]
@lheim It looks like you are using an older Version of Tensorflow . Many bugs have been fixed in the latest version. Could you please execute your code using Latest Version 2.4.1 or 2.5 and let us know if the issue still persists? Thanks!

**Comment by lheim** [comment]
@sushreebarsa I'm not working on this project anymore, so I can't test if the problem still exists. Nonetheless, feel free to close the issue if you think it's fixed now.
Cheers!

**Comment by sushreebarsa** [comment]
Can someone from the community can clarify if this issue is resolved so that we can close it ? Thanks!
